BUG: A test_to_datetime.py test fails on Python 3.15

### Pandas version checks

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [x] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.


### Reproducible Example

```python
The following test (pytest -k test_to_datetime_parse_timezone_malformed) fails with 3.0.5 when building with 3.15 (on Fedora 45/rawhide) and segfaults.


pandas/tests/tools/test_to_datetime.py::TestTimeConversionFormats::test_to_datetime_parse_timezone_malformed[]


The crash:

Fatal Python error: Segmentation fault
Thread 0x00007f12801eb6c0 [ThreadPoolExecu] (most recent call first):
  File "/usr/lib64/python3.15/concurrent/futures/thread.py", line 116 in _worker
  File "/usr/lib64/python3.15/threading.py", line 1160 in run
  File "/usr/lib64/python3.15/threading.py", line 1218 in _bootstrap_inner
  File "/usr/lib64/python3.15/threading.py", line 1180 in _bootstrap
Thread 0x00007f14719ff6c0 [python3] (most recent call first):
  File "/usr/lib/python3.15/site-packages/execnet/gateway_base.py", line 534 in read
  ...
Current thread 0x00007f14732d5180 [python3] (most recent call first):
  File ".../pandas/core/tools/datetimes.py", line 470 in _array_strptime_with_fallback
  File ".../pandas/core/tools/datetimes.py", line 435 in _convert_listlike_datetimes
  File ".../pandas/core/tools/datetimes.py", line 1072 in to_datetime
  File ".../pandas/tests/tools/test_to_datetime.py", line 535 in test_to_datetime_parse_timezone_malformed
  ...
Current thread's C stack trace (most recent call first):
  Binary file "/lib64/libpython3.15.so.1.0", at _Py_DumpStack+0x4c [0x7f1472d96d88]
  Binary file "/lib64/libpython3.15.so.1.0", at +0x3a0289 [0x7f1472da0289]
  Binary file "/lib64/libc.so.6", at +0x94070 [0x7f1472898070]
  Binary file ".../pandas/_libs/tslibs/strptime.cpython-315-x86_64-linux-gnu.so", at +0x23ca5 [0x7f1320d1bca5]
  Binary file ".../pandas/_libs/tslibs/strptime.cpython-315-x86_64-linux-gnu.so", at +0x2826f [0x7f1320d2026f]
  Binary file ".../pandas/_libs/tslibs/strptime.cpython-315-x86_64-linux-gnu.so", at +0x1a9f9 [0x7f1320d129f9]
  Binary file ".../pandas/_libs/tslibs/strptime.cpython-315-x86_64-linux-gnu.so", at +0x2045a [0x7f1320d1845a]
  ... (truncated, remainder is CPython call-machinery frames)
[gw0] node down: Not properly terminated
[gw0] FAILED .../test_to_datetime.py::TestTimeConversionFormats::test_to_datetime_parse_timezone_malformed[]
replacing crashed worker gw0
```

### Issue Description

The test appears to iterate over a parametrized over a list of deliberately malformed offset strings to be able to handle unintended input, I would expect to get an error in response and not a crash. The failing test appears to be when offset="", the constructed date string is "2010-01-01 12:00:00 " (trailing space, no offset).

### Expected Behavior

Pandas error out and not actually crash.

### Installed Versions

Pandas 3.0.5 with python 3.15-3.15.0~b4-1.fc45 on Fedora 45/rawhide

--- a/.github/workflows/unit-tests.yml
+++ b/.github/workflows/unit-tests.yml
@@ -443,7 +443,7 @@ jobs:
     #    To freeze this file, uncomment out the ``if: false`` condition, and migrate the jobs
     #    to the corresponding posix/windows-macos/sdist etc. workflows.
     # Feel free to modify this comment as necessary.
-    if: false
+    # if: false
     permissions:
       contents: read
     runs-on: ${{ matrix.os }}

--- a/.github/workflows/unit-tests.yml
+++ b/.github/workflows/unit-tests.yml
@@ -480,7 +480,10 @@ jobs:
           python -m pip install --upgrade pip wheel meson[ninja] meson-python
           python -m pip install --pre --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy
           python -m pip install versioneer[toml] python-dateutil tzdata cython hypothesis>=6.116.0 "pytest>=8.3.4" pytest-xdist>=3.6.1 pytest-cov
-          python -m pip install -ve . --no-build-isolation --no-index --no-deps -Csetup-args="--werror" -Csetup-args=-Dcpp_args="-Wno-missing-field-initializers"
+          python -m pip install -ve . --no-build-isolation --no-index --no-deps \
+            -Csetup-args="--werror" \
+            ${{ matrix.os != 'windows-2025' && '-Csetup-args=-Dcpp_args="-Wno-missing-field-initializers"' || '' }} \
+            ${{ matrix.os == 'windows-2025' && '-Csetup-args="--vsenv"' || '' }}
           python -m pip list
 
       - name: Run Tests

--- a/pandas/_libs/tslibs/strptime.pyx
+++ b/pandas/_libs/tslibs/strptime.pyx
@@ -181,7 +181,8 @@ cdef dict _parse_code_table = {"y": 0,
                                "G": 20,
                                "V": 21,
                                "u": 22,
-                               "N": 23}
+                               "N": 23,
+                               "colon_z": 24}
 
 
 cdef _validate_fmt(str fmt):
@@ -705,9 +706,19 @@ cdef tzinfo _parse_with_format(
                 f"time data \"{val}\" doesn't match format \"{fmt}\""
             )
         if len(val) != found.end():
+            rest = val[found.end():]
+            # Specific check for '%:z' directive
+            if (
+                "colon_z" in found.re.groupindex
+                and found.group("colon_z") is not None
+                and rest[0] != ":"
+            ):
+                raise ValueError(
+                    f"Missing colon in %:z before '{rest}', got '{val}'"
+                )
             raise ValueError(
                 "unconverted data remains when parsing with "
-                f"format \"{fmt}\": \"{val[found.end():]}\""
+                f"format \"{fmt}\": \"{rest}\""
             )
 
     else:
@@ -741,6 +752,7 @@ cdef tzinfo _parse_with_format(
         #      worthless without day of the week
         parse_code = _parse_code_table[group_key]
 
+        # if group_key == 'y':
         if parse_code == 0:
             year = int(group_val)
             # Open Group specification for strptime() states that a %y
@@ -752,9 +764,11 @@ cdef tzinfo _parse_with_format(
             else:
                 year += 1900
                 # TODO: not reached in tests 2023-10-28
+        # elif group_key == 'Y':
         elif parse_code == 1:
             # e.g. val='17-10-2010 07:15:30'; fmt='%d-%m-%Y %H:%M:%S'
             year = int(group_val)
+        # elif group_key == 'm':
         elif parse_code == 2:
             # e.g. val='17-10-2010 07:15:30'; fmt='%d-%m-%Y %H:%M:%S'
             month = int(group_val)
@@ -774,6 +788,7 @@ cdef tzinfo _parse_with_format(
         elif parse_code == 6:
             # e.g. val='17-10-2010 07:15:30'; fmt='%d-%m-%Y %H:%M:%S'
             hour = int(group_val)
+        # elif group_key == 'I':
         elif parse_code == 7:
             hour = int(group_val)
             ampm = found_dict.get("p", "").lower()
@@ -797,12 +812,15 @@ cdef tzinfo _parse_with_format(
                     hour += 12
                     # TODO: the implicit `else` branch is not tested 2023-10-28
             # TODO: the implicit `else` branch is not reached 2023-10-28; possible?
+        # elif group_key == 'M':
         elif parse_code == 8:
             # e.g. val='17-10-2010 07:15:30'; fmt='%d-%m-%Y %H:%M:%S'
             minute = int(group_val)
+        # elif group_key == 'S':
         elif parse_code == 9:
             # e.g. val='17-10-2010 07:15:30'; fmt='%d-%m-%Y %H:%M:%S'
             second = int(group_val)
+        # elif group_key == 'f':
         elif parse_code == 10:
             # e.g. val='10:10:10.100'; fmt='%H:%M:%S.%f'
             s = group_val
@@ -824,12 +842,15 @@ cdef tzinfo _parse_with_format(
             us = int(s)
             ns = us % 1000
             us = us // 1000
+        # elif group_key == 'A':
         elif parse_code == 11:
             # e.g val='Tuesday 24 Aug 2021 01:30:48 AM'; fmt='%A %d %b %Y %I:%M:%S %p'
             weekday = f_weekday_lookup[group_val.lower()]
+        # elif group_key == 'a':
         elif parse_code == 12:
             # e.g. val='Tue 24 Aug 2021 01:30:48 AM'; fmt='%a %d %b %Y %I:%M:%S %p'
             weekday = a_weekday_lookup[group_val.lower()]
+        # elif group_key == 'w':
         elif parse_code == 13:
             weekday = int(group_val)
             if weekday == 0:
@@ -838,9 +859,11 @@ cdef tzinfo _parse_with_format(
             else:
                 # e.g. val='2009324'; fmt='%Y%W%w'
                 weekday -= 1
+        # elif group_key == 'j':
         elif parse_code == 14:
             # e.g. val='2009164202000'; fmt='%Y%j%H%M%S'
             julian = int(group_val)
+        # elif group_key in ('U', 'W'):
         elif parse_code == 15 or parse_code == 16:
             week_of_year = int(group_val)
             if group_key == "U":
@@ -854,8 +877,14 @@ cdef tzinfo _parse_with_format(
         elif parse_code == 17:
             # e.g. val='2011-12-30T00:00:00.000000UTC'; fmt='%Y-%m-%dT%H:%M:%S.%f%Z'
             tz = zoneinfo.ZoneInfo(group_val)
-        elif parse_code == 19:
+        # elif group_key in ('z', 'colon_z'):
+        elif parse_code == 19 or parse_code == 24:
             # e.g. val='March 1, 2018 12:00:00+0400'; fmt='%B %d, %Y %H:%M:%S%z'
+            if group_val is None:
+                raise ValueError(
+                    f"time data \"{val}\" doesn't match format \"{fmt}\""
+                )
+
             tz = parse_timezone_directive(group_val)
         elif parse_code == 20:
             # e.g. val='2015-1-7'; fmt='%G-%V-%u'

--- a/pandas/compat/__init__.py
+++ b/pandas/compat/__init__.py
@@ -21,6 +21,7 @@ from pandas.compat._constants import (
     PY312,
     PY313,
     PY314,
+    PY315,
     PYPY,
     WASM,
 )
@@ -150,6 +151,7 @@ __all__ = [
     "PY312",
     "PY313",
     "PY314",
+    "PY315",
     "PYARROW_INSTALLED",
     "PYARROW_MIN_VERSION",
     "PYPY",

--- a/pandas/compat/_constants.py
+++ b/pandas/compat/_constants.py
@@ -16,6 +16,7 @@ IS64 = sys.maxsize > 2**32
 PY312 = sys.version_info >= (3, 12)
 PY313 = sys.version_info >= (3, 13)
 PY314 = sys.version_info >= (3, 14)
+PY315 = sys.version_info >= (3, 15)
 PYPY = platform.python_implementation() == "PyPy"
 WASM = (sys.platform == "emscripten") or (platform.machine() in ["wasm32", "wasm64"])
 ISMUSL = "musl" in (sysconfig.get_config_var("HOST_GNU_TYPE") or "")
@@ -32,6 +33,7 @@ __all__ = [
     "PY312",
     "PY313",
     "PY314",
+    "PY315",
     "PYPY",
     "WASM",
 ]

--- a/pandas/io/xml.py
+++ b/pandas/io/xml.py
@@ -342,44 +342,53 @@ class _XMLFrameParser:
             set(self.iterparse[row_node])
         )
 
-        for event, elem in iterparse(self.path_or_buffer, events=("start", "end")):
-            curr_elem = elem.tag.split("}")[1] if "}" in elem.tag else elem.tag
+        parser = iterparse(self.path_or_buffer, events=("start", "end"))
+        try:
+            for event, elem in parser:
+                curr_elem = elem.tag.split("}")[1] if "}" in elem.tag else elem.tag
 
-            if event == "start":
-                if curr_elem == row_node:
-                    row = {}
+                if event == "start":
+                    if curr_elem == row_node:
+                        row = {}
 
-            if row is not None:
-                if self.names and iterparse_repeats:
-                    for col, nm in zip(
-                        self.iterparse[row_node], self.names, strict=True
-                    ):
-                        if curr_elem == col:
-                            elem_val = elem.text if elem.text else None
-                            if elem_val not in row.values() and nm not in row:
-                                row[nm] = elem_val
+                if row is not None:
+                    if self.names and iterparse_repeats:
+                        for col, nm in zip(
+                            self.iterparse[row_node], self.names, strict=True
+                        ):
+                            if curr_elem == col:
+                                elem_val = elem.text if elem.text else None
+                                if elem_val not in row.values() and nm not in row:
+                                    row[nm] = elem_val
 
-                        if col in elem.attrib:
-                            if elem.attrib[col] not in row.values() and nm not in row:
-                                row[nm] = elem.attrib[col]
-                else:
-                    for col in self.iterparse[row_node]:
-                        if curr_elem == col:
-                            row[col] = elem.text if elem.text else None
-                        if col in elem.attrib:
-                            row[col] = elem.attrib[col]
+                            if col in elem.attrib:
+                                if (
+                                    elem.attrib[col] not in row.values()
+                                    and nm not in row
+                                ):
+                                    row[nm] = elem.attrib[col]
+                    else:
+                        for col in self.iterparse[row_node]:
+                            if curr_elem == col:
+                                row[col] = elem.text if elem.text else None
+                            if col in elem.attrib:
+                                row[col] = elem.attrib[col]
 
-            if event == "end":
-                if curr_elem == row_node and row is not None:
-                    dicts.append(row)
-                    row = None
+                if event == "end":
+                    if curr_elem == row_node and row is not None:
+                        dicts.append(row)
+                        row = None
 
-                elem.clear()
-                if hasattr(elem, "getprevious"):
-                    while (
-                        elem.getprevious() is not None and elem.getparent() is not None
-                    ):
-                        del elem.getparent()[0]
+                    elem.clear()
+                    if hasattr(elem, "getprevious"):
+                        while (
+                            elem.getprevious() is not None
+                            and elem.getparent() is not None
+                        ):
+                            del elem.getparent()[0]
+        finally:
+            if hasattr(parser, "close"):
+                parser.close()
 
         if dicts == []:
             raise ParserError("No result from selected items in iterparse.")

--- a/pandas/tests/extension/decimal/test_decimal.py
+++ b/pandas/tests/extension/decimal/test_decimal.py
@@ -163,7 +163,13 @@ class TestDecimalArray(base.ExtensionTests):
         # GH#57723
         # EAs that don't have special logic for None will raise, unlike pandas'
         # which interpret None as the NA value for the dtype.
-        msg = "conversion from NoneType to Decimal is not supported"
+        msg = "|".join(
+            [
+                "Cannot convert None to Decimal",  # PY315 - maybe linux specific
+                "conversion from NoneType to Decimal is not supported",
+            ]
+        )
+
         with pytest.raises(TypeError, match=msg):
             super().test_fillna_with_none(data_missing)
 

--- a/pandas/tests/frame/indexing/test_indexing.py
+++ b/pandas/tests/frame/indexing/test_indexing.py
@@ -10,6 +10,7 @@ import numpy as np
 import pytest
 
 from pandas._libs import iNaT
+from pandas.compat import PY315
 
 from pandas.core.dtypes.common import is_integer
 
@@ -29,7 +30,10 @@ from pandas import (
 import pandas._testing as tm
 
 # We pass through a TypeError raised by numpy
-_slice_msg = "slice indices must be integers or None or have an __index__ method"
+if PY315:
+    _slice_msg = "slice indices must be integers or have an __index__ method"
+else:
+    _slice_msg = "slice indices must be integers or None or have an __index__ method"
 
 
 class TestDataFrameIndexing:

--- a/pandas/tests/indexes/period/test_partial_slicing.py
+++ b/pandas/tests/indexes/period/test_partial_slicing.py
@@ -1,6 +1,8 @@
 import numpy as np
 import pytest
 
+from pandas.compat import PY315
+
 from pandas import (
     DataFrame,
     PeriodIndex,
@@ -53,7 +55,10 @@ class TestPeriodIndex:
         # GH#6716
         idx = make_range(start="2013/01/01", freq="D", periods=400)
 
-        msg = "slice indices must be integers or None or have an __index__ method"
+        if PY315:
+            msg = "slice indices must be integers or have an __index__ method"
+        else:
+            msg = "slice indices must be integers or None or have an __index__ method"
         # slices against index should raise IndexError
         values = [
             "2014",
@@ -82,7 +87,10 @@ class TestPeriodIndex:
     def test_range_slice_seconds(self, make_range):
         # GH#6716
         idx = make_range(start="2013/01/01 09:00:00", freq="s", periods=4000)
-        msg = "slice indices must be integers or None or have an __index__ method"
+        if PY315:
+            msg = "slice indices must be integers or have an __index__ method"
+        else:
+            msg = "slice indices must be integers or None or have an __index__ method"
 
         # slices against index should raise IndexError
         values = [

--- a/pandas/tests/indexes/test_old_base.py
+++ b/pandas/tests/indexes/test_old_base.py
@@ -7,6 +7,7 @@ import numpy as np
 import pytest
 
 from pandas._libs.tslibs import Timestamp
+from pandas.compat import PY315
 from pandas.errors import Pandas4Warning
 
 from pandas.core.dtypes.common import (
@@ -463,6 +464,8 @@ class TestBase:
         if len(index) == 0:
             # 0 vs 0.5 in error message varies with numpy version
             msg = "index (0|0.5) is out of bounds for axis 0 with size 0"
+        elif PY315:
+            msg = "slice indices must be integers or have an __index__ method"
         else:
             msg = "slice indices must be integers or None or have an __index__ method"
 

--- a/pandas/tests/indexing/test_floats.py
+++ b/pandas/tests/indexing/test_floats.py
@@ -1,6 +1,8 @@
 import numpy as np
 import pytest
 
+from pandas.compat import PY315
+
 from pandas import (
     DataFrame,
     Index,
@@ -251,7 +253,13 @@ class TestFloatIndexers:
         # setitem
         if indexer_sli is tm.iloc:
             # otherwise we keep the same message as above
-            msg = "slice indices must be integers or None or have an __index__ method"
+            if PY315:
+                msg = "slice indices must be integers or have an __index__ method"
+            else:
+                msg = (
+                    "slice indices must be integers or None or have an __index__ method"
+                )
+
         with pytest.raises(TypeError, match=msg):
             indexer_sli(s)[idx] = 0
 

--- a/pandas/tests/scalar/timestamp/test_constructors.py
+++ b/pandas/tests/scalar/timestamp/test_constructors.py
@@ -23,6 +23,7 @@ from pandas._libs.tslibs.dtypes import NpyDatetimeUnit
 from pandas.compat import (
     PY313,
     PY314,
+    PY315,
 )
 from pandas.errors import (
     OutOfBoundsDatetime,
@@ -268,7 +269,11 @@ class TestTimestampConstructorPositionalAndKeywordSupport:
     def test_constructor_keyword(self):
         # GH#10758
         msg = "|".join(
-            ["function missing required argument 'day'", "Required argument 'day'"]
+            [
+                r"datetime\(\) missing required argument 'day'",  # PY315
+                "function missing required argument 'day'",
+                "Required argument 'day'",
+            ]
         )
         with pytest.raises(TypeError, match=msg):
             Timestamp(year=2000, month=1)
@@ -449,7 +454,15 @@ class TestTimestampConstructorPositionalAndKeywordSupport:
         # GH#31200
 
         # The exact error message of datetime() depends on its version
-        msg1 = r"function missing required argument '(year|month|day)' \(pos [123]\)"
+        if PY315:
+            msg1 = (
+                r"datetime\(\) missing required argument "
+                r"'(year|month|day)' \(pos [123]\)"
+            )
+        else:
+            msg1 = (
+                r"function missing required argument '(year|month|day)' \(pos [123]\)"
+            )
         msg2 = r"Required argument '(year|month|day)' \(pos [123]\) not found"
         msg = "|".join([msg1, msg2])
 

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -535,6 +535,39 @@ class TestTimeConversionFormats:
         expected = DatetimeIndex(expected_dates)
         tm.assert_index_equal(result, expected)
 
+    @pytest.mark.xfail(
+        not PY315, reason="%:z directive not supported prior to 3.15", raises=ValueError
+    )
+    def test_to_datetime_colon_z_offset(self):
+        dates = [
+            "2010-01-01 12:00:00+04:00",
+            "2010-01-01 12:00:00+04:30",
+            "2010-01-01 12:00:00-05:00",
+        ]
+        expected_dates = [
+            "2010-01-01 08:00:00+00:00",
+            "2010-01-01 07:30:00+00:00",
+            "2010-01-01 17:00:00+00:00",
+        ]
+        fmt = "%Y-%m-%d %H:%M:%S%:z"
+
+        result = to_datetime(dates, format=fmt, utc=True)
+        expected = DatetimeIndex(expected_dates)
+        tm.assert_index_equal(result, expected)
+
+    def test_to_datetime_missing_colon_z_offset(self):
+        # test adapted from python/cpython#136961
+        dates = ["+04:0030"]
+        fmt = "%:z"
+
+        if PY315:
+            msg = r"Missing colon in %:z before '30', got '\+04:0030'"
+        else:
+            msg = "':' is a bad directive in format '%:z'"
+
+        with pytest.raises(ValueError, match=msg):
+            to_datetime(dates, format=fmt, utc=True)
+
     @pytest.mark.parametrize(
         "offset", ["+0", "-1foo", "UTCbar", ":10", "+01:000:01", ""]
     )

--- a/pandas/tests/tools/test_to_datetime.py
+++ b/pandas/tests/tools/test_to_datetime.py
@@ -25,6 +25,7 @@ from pandas._libs.tslibs import (
 )
 from pandas.compat import (
     PY314,
+    PY315,
     WASM,
 )
 from pandas.errors import (
@@ -1447,9 +1448,12 @@ class TestToDatetime:
     @pytest.mark.parametrize("errors", ["coerce", "raise"])
     def test_invalid_format_raises(self, errors):
         # https://github.com/pandas-dev/pandas/issues/50255
-        with pytest.raises(
-            ValueError, match="':' is a bad directive in format 'H%:M%:S%"
-        ):
+        if PY315:
+            msg = r"':M' is a bad directive in format 'H%:M%:S%"
+        else:
+            msg = "':' is a bad directive in format 'H%:M%:S%"
+
+        with pytest.raises(ValueError, match=msg):
             to_datetime(["00:00:00"], format="H%:M%:S%", errors=errors)
 
     @pytest.mark.parametrize("value", ["a", "00:01:99"])

